workspaces/mcp-chat-client/plugins/mcp-chat-client-backend/src/services/MCPClientServiceImpl.ts

```diff
@@ -0,0 +1,349 @@
+/*
+ * Copyright 2025 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {
+  LoggerService,
+  RootConfigService,
+} from '@backstage/backend-plugin-api';
+import { Client } from '@modelcontextprotocol/sdk/client/index.js';
+import { StreamableHTTPClientTransport } from '@modelcontextprotocol/sdk/client/streamableHttp.js';
+import { StdioClientTransport } from '@modelcontextprotocol/sdk/client/stdio.js';
+import { SSEClientTransport } from '@modelcontextprotocol/sdk/client/sse.js';
+import * as path from 'path';
+import {
+  ProviderFactory,
+  getProviderConfig as getConfig,
+  getProviderInfo,
+} from '../providers/provider-factory';
+import { ServerConfig } from '../types';
+import { executeToolCall, findNpxPath } from '../utils';
+import {
+  ChatMessage,
+  LLMProvider,
+  Tool,
+  ToolCall,
+} from '../providers/base-provider';
+import { MCPClientService } from './MCPClientService';
+
+export type Options = {
+  logger: LoggerService;
+  config: RootConfigService;
+};
+
+export class MCPClientServiceImpl implements MCPClientService {
+  private readonly logger: LoggerService;
+  private readonly config: RootConfigService;
+  private llmProvider!: LLMProvider;
```

Comment:
perhaps `initializeLLMProvider()` could return the llmProvider so that we remove the non null assertion here